kythe/docs/rfc/2909.md

```diff
@@ -0,0 +1,145 @@
+
+The scope of this proposal only includes FileDecorations and CrossReferences.  Other services could be made to use similar formats, but the most improvement will be gained from these two APIs.
+
+# Implementation
+
+Essentially each field of an existing ProtocolBuffer value in the serving data becomes zero or more key-value entries depending on whether it's unset, set, or a repeated field.  The bulk of the field data will be represented in the key and the keys will be encoded as [orderedcodes](https://godoc.org/github.com/google/orderedcode).  It's important that each key be unique to fit storage models like LevelDB which are not multimaps (unlike the underlying SSTables).  Tags within keys will indicate how to decode the values themselves and help order the key-value entries.  The ordering of the keys will be chosen as to best fit a performant server, allowing for easy, single-pass scans to retrieve all possible data.
+
+## FileDecorations
+
+### Existing ProtocolBuffer definition:
+
+```protobuf
+// Simplified FileDecorations
+message FileDecorations {
+  File file = 1;
+
+  repeated Decoration decoration = 2;
+  repeated ExpandedAnchor target_definitions = 3;
+  repeated Node target = 4;
+  repeated Override target_override = 5;
+  repeated kythe.proto.common.Diagnostic diagnostic = 6;
+
+  message Decoration {
+    RawAnchor anchor = 1;
+    string kind = 2;
+    string target = 5;
+    string target_definition = 4;
+  }
+  message Override {
+    enum Kind {
+      OVERRIDES  = 0;
+      EXTENDS    = 1;
+    }
+    string overriding = 1;
+    string overridden = 2;
+    string overridden_definition = 5;
+    Kind kind = 3;
+    kythe.proto.common.MarkedSource marked_source = 4;
+  }
+}
+```
+
+### Key-value format:
+
+| key | value | kind |
+|---|---|---|
+| `"fd"-file` | `FileDecorationsIndex` | file metadata |
+| `"fd"-file-00-start-end` | bytes | file contents for span |
+| `"fd"-file-10-start-end-kind-target` | `<empty>` | decoration targets |
+| `"fd"-file-20-target-kind-override` | `<empty>` | override per target |
+| `"fd"-file-30-target` | `srvpb.Node` | target node facts |
+| `"fd"-file-40-target` | `spb.VName` | definition for each target |
+| `"fd"-file-50-def` | `srvpb.ExpandedAnchor` | each definition's location |
+| `"fd"-file-60-override` | `MarkedSource` | `MarkedSource` per override |
+| `"fd"-file-70-start-end-hash` | `Diagnostic` | `Diagnostic` per span (use `-1:-1` span for entire file) |
+
+
+Notes:
+
+* Each value will be contained within a wrapper message to support evolution of the table format
+* All file decoration entries are prefixed by "fd" and the file's `VName`
```

Comment:
I'm curious why the quoting?

Reply:
The quoting is just notation to indicate a literal string.